### PR TITLE
Validator deactivation and reactivation

### LINKED ISSUE #35: PoS validator deactivation + reactivation
add transactions and update PoS validity predicate for validator deactivation + reactivation

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201757896515792/1201763046898531) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by brentstone**
## Describe your changes
Closes #35.

## Indicate on which release or other PRs this topic is based on
v0.25.0

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: wasm/wasm_source/src/tx_deactivate_validator.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//! A tx to deactivate a validator.
```
Comment (tzemanovic) [inline wasm/wasm_source/src/tx_deactivate_validator.rs:1]: we could use a single tx type that takes a bool for this, but it's not much difference

File: proof_of_stake/src/lib.rs [hunk 2]
```diff
@@ -5335,6 +5323,198 @@ where
...
+
+    // Set the state to inactive
+    validator_state_handle(validator).set(
+        storage,
+        ValidatorState::Inactive,
+        current_epoch,
+        params.pipeline_len,
+    )?;
+
+    // TODO: what other kinds of data should be deleted, if any??
```
Comment (tzemanovic) [inline proof_of_stake/src/lib.rs:5444]: I think validator set + position is sufficient

File: wasm/wasm_source/src/tx_deactivate_validator.rs (hunk 1 above)
Comment (brentstone) [reply wasm/wasm_source/src/tx_deactivate_validator.rs:1]: Ah true, but unless you think there's a nice advantage, I'd rather take the path of least resistance rn and keep it as is.

File: proof_of_stake/src/lib.rs [hunk 3]
```diff
@@ -5335,6 +5323,198 @@ where
...
+pub fn reactivate_validator<S>(
+    storage: &mut S,
+    validator: &Address,
+    current_epoch: Epoch,
+) -> storage_api::Result<()>
+where
+    S: StorageRead + StorageWrite,
+{
+    // TODO: need to additionally check for past slashes, in case a jailed
+    // validator tries to deactivate and then reactivate
```
Comment (tzemanovic) [inline proof_of_stake/src/lib.rs:5459]: could we prevent deactivation on a jailed validator for this instead?

Comment (brentstone) [reply proof_of_stake/src/lib.rs:5459]: Yeah I suppose this is easier and logical too. Will do.

Comment (brentstone) [reply proof_of_stake/src/lib.rs:5464]: Done in 909afde607e046317958839eb70efbdf457ca1ec

**Review by tzemanovic — APPROVED** [review]

**Pull request merged**